packages/customize-widgets/src/components/sidebar-block-editor/use-sidebar-block-editor.js

```diff
@@ -160,107 +158,105 @@ export default function useSidebarBlockEditor( sidebar ) {
 
 					setState( ( lastState ) => {
 						const blocksByWidgetId = keyBy(
-							lastState.blocks,
+							lastState,
 							'attributes.__internalWidgetId'
 						);
 
-						return {
-							...lastState,
-							blocks: widgetIds.map(
-								( widgetId ) => blocksByWidgetId[ widgetId ]
-							),
-						};
+						return widgetIds.map(
+							( widgetId ) => blocksByWidgetId[ widgetId ]
+						);
 					} );
 					break;
 			}
 		};
 
-		sidebar.subscribe( handler );
-		return () => sidebar.unsubscribe( handler );
+		return sidebar.subscribe( handler );
 	}, [ sidebar ] );
 
 	const onChangeBlocks = useCallback(
 		( _nextBlocks ) => {
 			ignoreIncoming.current = true;
 
-			const blocksByWidgetId = keyBy(
-				state.blocks,
-				( block ) => block.attributes.__internalWidgetId
-			);
+			setState( ( lastState ) => {
+				const blocksByWidgetId = keyBy(
+					lastState,
+					( block ) => block.attributes.__internalWidgetId
+				);
 
-			const nextBlocks = _nextBlocks.map( ( nextBlock, index ) => {
-				if (
-					nextBlock.attributes.__internalWidgetId &&
-					nextBlock.attributes.__internalWidgetId in blocksByWidgetId
-				) {
-					const block =
-						blocksByWidgetId[
-							nextBlock.attributes.__internalWidgetId
-						];
-					if ( ! isEqual( block, nextBlock ) ) {
-						const widgetId =
-							nextBlock.attributes.__internalWidgetId;
-						const widgetToUpdate = sidebar.getWidget( widgetId );
-						const widget = blockToWidget(
-							nextBlock,
-							widgetToUpdate
-						);
-						sidebar.updateWidget( widget );
+				const nextBlocks = _nextBlocks.map( ( nextBlock, index ) => {
+					if (
+						nextBlock.attributes.__internalWidgetId &&
```

Comment:
minor thing - `&&` is unnecessary here as long as we don't want to skip a case when an id is 0. If nextBlock.attributes.__internalWidgetId is `undefined` then `undefined in {}` will return `false`.

Reply:
This is actually needed though. If there's no `__internalWidgetId` then the line above creating `blocksByWidgetId` will includes a `undefined` key. Then `undefined in blocksByWidgetId` will return true. We want to filter out that case.